### ISSUE TITLE
ruby/rubygems & certificates updated, automatic ssl check keeps failing

### ISSUE THREAD

**Issue by WattnDing** [rubygems/rubygems#2632]
**Disclaimer**: This looks just like another edition of the SSL connection problems that have piled up in your issue tracker over the years. However, I am running dry now, after two days of checks and read-ups and no success, so far. I fear that you will have to find a flaw in my procedures, rather than the original source of mischief... 

This issue is related to:
  - [x] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

**Concerns** gem update/install ( Errno::ECONNRESET: Connection reset by peer)
I do not use RVM nor other Ruby package managers.

**The automatic ssl-check** (`ruby -ropen-uri -e 'eval open("https://git.io/vQhWq").read'`)
```
Traceback (most recent call last):
	13: from -e:1:in `<main>'
	12: from /usr/lib/ruby/2.6.0/open-uri.rb:35:in `open'
	11: from /usr/lib/ruby/2.6.0/open-uri.rb:736:in `open'
	10: from /usr/lib/ruby/2.6.0/open-uri.rb:165:in `open_uri'
	 9: from /usr/lib/ruby/2.6.0/open-uri.rb:224:in `open_loop'
	 8: from /usr/lib/ruby/2.6.0/open-uri.rb:224:in `catch'
	 7: from /usr/lib/ruby/2.6.0/open-uri.rb:226:in `block in open_loop'
	 6: from /usr/lib/ruby/2.6.0/open-uri.rb:756:in `buffer_open'
	 5: from /usr/lib/ruby/2.6.0/open-uri.rb:337:in `open_http'
	 4: from /usr/lib/ruby/2.6.0/net/http.rb:919:in `start'
	 3: from /usr/lib/ruby/2.6.0/net/http.rb:930:in `do_start'
	 2: from /usr/lib/ruby/2.6.0/net/http.rb:996:in `connect'
	 1: from /usr/lib/ruby/2.6.0/net/protocol.rb:44:in `ssl_socket_connect'
/usr/lib/ruby/2.6.0/net/protocol.rb:44:in `connect_nonblock': SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate) (OpenSSL::SSL::SSLError)
```

**The guide** (https://bundler.io/v1.16/guides/rubygems_tls_ssl_troubleshooting_guide.html)  I have read, the rubygems package I have updated (successively, once for each Ruby-version, the current one inclusive), the GlobalSignRootCA.pem I have replaced.

**Immediate question**: I mentioned that I have followed the guide, but you can ask a pertinent question and I will tell you, where and which certificate I have... what can go wrong, you ask. And I do not know the answer.

**Current RubyGems Environment** (Ruby had been 2.6.0 from snapshot or 2.7.x from git with identical failures):
  - RUBYGEMS VERSION: 3.0.2
  - RUBY VERSION: 2.5.3 (2018-10-18 patchlevel 105) [x86_64-linux-gnu]
  - INSTALLATION DIRECTORY: /var/lib/gems/2.5.0
  - USER INSTALLATION DIRECTORY: /home/[user]/.gem/ruby/2.5.0
  - RUBY EXECUTABLE: /usr/bin/ruby2.5
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /usr/local/bin
  - SPEC CACHE DIRECTORY: /home/[user]/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /var/lib/gems/2.5.0
     - /home/[user]/.gem/ruby/2.5.0
     - /usr/lib/x86_64-linux-gnu/rubygems-integration/2.5.0
     - /usr/share/rubygems-integration/2.5.0
     - /usr/share/rubygems-integration/all
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/local/games
     - /usr/games
     - /home/[user]/bin

**Bundler environment** (I have never used bundler before, this is the first time I run it: bundler env
```
Bundler       2.0.1
  Platforms   ruby, x86_64-linux
Ruby          2.5.3p105 (2018-10-18 revision 65156) [x86_64-linux-gnu]
  Full Path   /usr/bin/ruby2.5
  Config Dir  /etc
RubyGems      3.0.2
  Gem Home    /var/lib/gems/2.5.0
  Gem Path    /home/michael/.gem/ruby/2.5.0:/var/lib/gems/2.5.0:/usr/lib/x86_64-linux-gnu/rubygems-integration/2.5.0:/usr/share/rubygems-integration/2.5.0:/usr/share/rubygems-integration/all
  User Path   /home/michael/.gem/ruby/2.5.0
  Bin Dir     /usr/local/bin
Tools         
  Git         2.20.1
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2019-01-04
Git SHA           d7ad2192f
Released Version  true
```
**OpenSSL Library** (coming from Debian Buster. OpenSSL from git sources does not make a difference)
OpenSSL 1.1.2-dev  xx XXX xxxx

**Operating System** is Debian GNU/Linux buster/sid

**Really** I bet the solution is hidden in the docs or your past discussions, but I cannot see it. My English may fail me, too :(

**Comment by WattnDing** [comment]
I give up now and download gems via https from the Web-Site. **Ω**

**Comment by deivid-rodriguez** [comment]
Sorry your issue wasn't attended before... :(

There's a funny thing about our SSL checker. Don't you need SSL setup properly to actually run the script in the recommeded way? It sounds like the error you're getting. Can you download the script to a ".rb" file and run it (`ruby <name_of_the_file.rb>`)?

**Comment by WattnDing** [comment]
Hi Deivid
> There's a funny thing about our SSL checker. 
 Which ?

> Don't you need SSL setup properly to actually run the script in the recommeded way? 
I can say “I guess so” in response to half your phrase. Which script are you talking about?

> It sounds like the error you're getting. Can you download the script to a ".rb" file and run it (`ruby 
> <name_of_the_file.rb>`)?

Now I am at a loss. Is it possible that you misunderstood my OP? All I want (for now) is be able to update the installed ruby-gems ( user@machine:~$ sudo gem update).
If that is not, what you are responding to, I am sorry for the confusion. But anyway, I deem it impossible, for the time, to get rid of the trouble anytime soon and content with using the Web (as it used to be in the good old days). Maybe we should all just return to FTP.

CLOFI.

**Comment by deivid-rodriguez** [comment]
I'm sorry, I should've explained myself better. I meant this part of your original post:

> **The automatic ssl-check** (`ruby -ropen-uri -e 'eval open("https://git.io/vQhWq").read'`)
> ```
> Traceback (most recent call last):
>	13: from -e:1:in `<main>'
>	12: from /usr/lib/ruby/2.6.0/open-uri.rb:35:in `open'
>	11: from /usr/lib/ruby/2.6.0/open-uri.rb:736:in `open'
>	10: from /usr/lib/ruby/2.6.0/open-uri.rb:165:in `open_uri'
>	 9: from /usr/lib/ruby/2.6.0/open-uri.rb:224:in `open_loop'
>	 8: from /usr/lib/ruby/2.6.0/open-uri.rb:224:in `catch'
>	 7: from /usr/lib/ruby/2.6.0/open-uri.rb:226:in `block in open_loop'
>	 6: from /usr/lib/ruby/2.6.0/open-uri.rb:756:in `buffer_open'
>	 5: from /usr/lib/ruby/2.6.0/open-uri.rb:337:in `open_http'
>	 4: from /usr/lib/ruby/2.6.0/net/http.rb:919:in `start'
>	 3: from /usr/lib/ruby/2.6.0/net/http.rb:930:in `do_start'
>	 2: from /usr/lib/ruby/2.6.0/net/http.rb:996:in `connect'
>	 1: from /usr/lib/ruby/2.6.0/net/protocol.rb:44:in `ssl_socket_connect'
>/usr/lib/ruby/2.6.0/net/protocol.rb:44:in `connect_nonblock': SSL_connect returned=1 errno=0  state=error: certificate verify failed (unable to get local issuer certificate) (OpenSSL::SSL::SSLError)
> ```

That's part of our troubleshooting steps, and it's crashing for you. That's because you actually need SSL to run it! So I was recommending that you visit that url (since I assume your browser is working), download and save the contents of that script to a ruby file, and run that file.

**Comment by WattnDing** [comment]
Here are the results of the jury:
------------
```
user@machine:/tmp$ ./ssl_check.rb 
Here's your Ruby and OpenSSL environment:

Ruby:           2.5.3p105 (2018-10-18 revision 65156) [x86_64-linux-gnu]
RubyGems:       3.0.2
Bundler:        2.0.1
Compiled with:  OpenSSL 1.1.1a  20 Nov 2018
Loaded version: OpenSSL 1.1.2-dev  xx XXX xxxx
SSL_CERT_FILE:  /usr/local/ssl/cert.pem
SSL_CERT_DIR:   /usr/local/ssl/certs

With that out of the way, let's see if you can connect to rubygems.org...

Bundler connection to rubygems.org:       failed  ❌  (Connection reset by peer - SSL_connect)
RubyGems connection to rubygems.org:      failed  ❌  (Errno::ECONNRESET: Connection reset by peer - SSL_connect (https://rubygems.org))
Ruby net/http connection to rubygems.org: failed  ❌

Unfortunately, this Ruby can't connect to rubygems.org. 😡
Your Ruby can't connect to rubygems.org because you are missing the certificate files OpenSSL needs to verify you are connecting to the genuine rubygems.org servers.
```
And here are the certificates, I know of:
```
user@machine:/tmp$ locate GlobalSignRootCA
/home/michael/prog/foreign/ruby/lib/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/home/michael/prog/foreign/ruby/lib/rubygems/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/home/michael/prog/foreign/ruby_git/lib/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/home/michael/prog/foreign/ruby_git/lib/rubygems/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/2.5.0/rubygems/ssl_certs/GlobalSignRootCA.pem
/usr/lib/ruby/2.6.0/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/2.6.0/rubygems/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/2.7.0/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/2.7.0/rubygems/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/gems/2.7.0/gems/rubygems-update-3.0.2/bundler/lib/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/gems/2.7.0/gems/rubygems-update-3.0.2/lib/rubygems/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/site_ruby/2.6.0/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/site_ruby/2.6.0/rubygems/ssl_certs/GlobalSignRootCA.pem
/usr/lib/ruby/site_ruby/2.6.0/rubygems/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/site_ruby/2.7.0/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/lib/ruby/site_ruby/2.7.0/rubygems/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/local/lib/site_ruby/2.5.0/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/usr/local/lib/site_ruby/2.5.0/rubygems/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem
/var/lib/gems/2.5.0/gems/bundler-2.0.1/lib/bundler/ssl_certs/index.rubygems.org/GlobalSignRootCA.pem

```

**Comment by MSP-Greg** [comment]
@WattnDing 

You're using OpenSSL 1.1.2-dev, and a lot of their recent CI is red.

Have you tested it?  Maybe with something like:

```
openssl s_client -connect www.google.com:443
```

compared to:
```
openssl s_client -connect repo.msys2.org:443
```

I've built & tested OpenSSL pre-releases before, but I usually wait until a bit closer to release.  Hence, I haven't worked with 1.1.2, either stand-alone or with Ruby.

**Comment by WattnDing** [comment]
This version 1.1.x is what I get from the Debian repositories. It is the most mainstream release that I can get right now.
But I have compiled a newer OpenSSL 3.0, as I used to before my troubles with the gem-utility. The command-line **openssl s_client -connect www.google.com:443**
produces
```
CONNECTED(00000004)
depth=1 C = US, O = Google Trust Services, CN = Google Internet Authority G3
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = California, L = Mountain View, O = Google LLC, CN = www.google.com
verify return:1
---
Certificate chain
 0 s:C = US, ST = California, L = Mountain View, O = Google LLC, CN = www.google.com
   i:C = US, O = Google Trust Services, CN = Google Internet Authority G3
 1 s:C = US, O = Google Trust Services, CN = Google Internet Authority G3
   i:OU = GlobalSign Root CA - R2, O = GlobalSign, CN = GlobalSign
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDxzCCAq+gAwIBAgIIcmtoyHPGf7IwDQYJKoZIhvcNAQELBQAwVDELMAkGA1UE
BhMCVVMxHjAcBgNVBAoTFUdvb2dsZSBUcnVzdCBTZXJ2aWNlczElMCMGA1UEAxMc
R29vZ2xlIEludGVybmV0IEF1dGhvcml0eSBHMzAeFw0xOTAxMjkxNDU4MDBaFw0x
OTA0MjMxNDU4MDBaMGgxCzAJBgNVBAYTAlVTMRMwEQYDVQQIDApDYWxpZm9ybmlh
MRYwFAYDVQQHDA1Nb3VudGFpbiBWaWV3MRMwEQYDVQQKDApHb29nbGUgTExDMRcw
FQYDVQQDDA53d3cuZ29vZ2xlLmNvbTBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IA
BBSwVB3jQpvxEACaUMGuRLoxAdRNv4CZdugnsLePx6XYCxfOi+CktJ/Z6liRbTaD
htXQb02KvRVd+wHthxS/vfCjggFSMIIBTjATBgNVHSUEDDAKBggrBgEFBQcDATAO
BgNVHQ8BAf8EBAMCB4AwGQYDVR0RBBIwEIIOd3d3Lmdvb2dsZS5jb20waAYIKwYB
BQUHAQEEXDBaMC0GCCsGAQUFBzAChiFodHRwOi8vcGtpLmdvb2cvZ3NyMi9HVFNH
SUFHMy5jcnQwKQYIKwYBBQUHMAGGHWh0dHA6Ly9vY3NwLnBraS5nb29nL0dUU0dJ
QUczMB0GA1UdDgQWBBTrALW2otfVyzLoqKYjitly6QXueTAMBgNVHRMBAf8EAjAA
MB8GA1UdIwQYMBaAFHfCuFCaZ3Z2sS3ChtCDoH6mfrpLMCEGA1UdIAQaMBgwDAYK
KwYBBAHWeQIFAzAIBgZngQwBAgIwMQYDVR0fBCowKDAmoCSgIoYgaHR0cDovL2Ny
bC5wa2kuZ29vZy9HVFNHSUFHMy5jcmwwDQYJKoZIhvcNAQELBQADggEBAIwCuzbB
5AWDiw6ld7EArSmoytsaK6dr9JmMbNo068niOTHUhYwuv6/VloOCjny86R8hG70w
WkPwAL89rUriqRLjUf/65Pe1Kj90N7JdUG22K1D6QnyIKm8F1lrbVGBPju5bfas7
yk3q6s7M2emwKU7XK8SZq41btl4TPCywo+TXGYWBJntSwM7Qq6r6vzAmkuO2fRtO
mjAK++qAQ3BjrXYe3CwbwSIF76Frwff9VOXP1yb/4ouT5Na13qbCc0FXHP/xCXoP
PWOVt566yut/kWVpda/UmpIm/YVo+z/8iLXx/7JK94P7uluS802w8rQBTPpokT2i
lr2/6PTD307l2JU=
-----END CERTIFICATE-----
subject=C = US, ST = California, L = Mountain View, O = Google LLC, CN = www.google.com

issuer=C = US, O = Google Trust Services, CN = Google Internet Authority G3

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: ECDSA
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 2582 bytes and written 401 bytes
Verification error: unable to get local issuer certificate
---
New, TLSv1.2, Cipher is ECDHE-ECDSA-CHACHA20-POLY1305
Server public key is 256 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-ECDSA-CHACHA20-POLY1305
    Session-ID: 140CE82932EBE1FC701D4194A0C60352D4B52FA26E3F3ED99A4E924603EF466D
    Session-ID-ctx: 
    Master-Key: BB75BEBA4D3F2DEBDFF387F7D3C54E493198FB6FE5661C6B8A7F157C4DE6A46CAC063EAC5324EF4F278C170A6949B6C1
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 100800 (seconds)
    TLS session ticket:
    0000 - 00 48 de 30 b4 4d fc 37-5a 34 2d 9d 13 45 3b d1   .H.0.M.7Z4-..E;.
    0010 - df b4 1b 58 bd 12 c5 2f-ee d5 51 c2 af 1d 47 5c   ...X.../..Q...G\
    0020 - 84 6c 09 fd c9 24 12 c3-f0 8f 75 3b bc 2b 03 09   .l...$....u;.+..
    0030 - 96 98 7f 11 d3 1b 13 a3-e3 d4 5f 22 94 b8 66 31   .........._"..f1
    0040 - 6c e1 f4 eb a8 34 7f f5-1e 99 e6 1b 7a 8b ae 97   l....4......z...
    0050 - b4 9f b4 27 5b f9 f0 72-37 b2 f6 ee 71 2d 4d ef   ...'[..r7...q-M.
    0060 - cc 24 ea ad 44 53 a5 84-9a ba 43 72 49 97 ab eb   .$..DS....CrI...
    0070 - c5 1c 40 de a5 db 54 52-59 8a 49 dc 50 e6 e4 a9   ..@...TRY.I.P...
    0080 - f5 71 aa a3 09 8a 7c d7-ec 10 60 65 09 ce d2 ef   .q....|...`e....
    0090 - e4 09 8d c7 57 d8 37 fa-f3 5b 30 a7 a6 1a 43 2e   ....W.7..[0...C.
    00a0 - 2a a8 64 91 3d d9 9f 39-c6 37 31 c0 6f 65 92 f9   *.d.=..9.71.oe..
    00b0 - 25 e7 b1 9e cd 9d f2 38-af 23 b0 18 b7 8e a8 ca   %......8.#......
    00c0 - 58 8f 3b df b4 f0 19 9a-06 f1 d9 61 78 66 18 21   X.;........axf.!
    00d0 - 64 b2 da a7 11 dd 1e 9b-fd 8e                     d.........

    Start Time: 1550427064
    Timeout   : 7200 (sec)
    Verify return code: 20 (unable to get local issuer certificate)
    Extended master secret: yes
---
read:errno=0
```

In contrast, with **openssl s_client -connect repo.msys2.org:443**
I get
```
140444383080704:error:0200206F:system library:connect:Connection refused:crypto/bio/b_sock2.c:110:
140444383080704:error:2008A067:BIO routines:BIO_connect:connect error:crypto/bio/b_sock2.c:111:
connect:errno=111

```

**Comment by MSP-Greg** [comment]
@WattnDing

> However, just for completeness in response to MSP-Greg

Thanks.  What I mentioned is due to the below, which means that Ruby's OpenSSL std-lib was compiled with OpenSSL 1.1.1a, but is loading files built with OpenSSL 1.1.2-dev.
```
Compiled with:  OpenSSL 1.1.1a  20 Nov 2018
Loaded version: OpenSSL 1.1.2-dev  xx XXX xxxx
```

From what you posted, OpenSSL 1.1.2-dev is working, but whether Ruby OpenSSL will operate correctly with it is another issue...

**Comment by WattnDing** [comment]
Could the following give an indication.., I will put a precise question further below.

Having had difficulties to fetch mail via SSL, I changed a setting which points at the directory with the diverse certificates from the ca-certificates package. As, before, I had used my own directory for the few certificates that I really need, but keeping them up-to-date appears to be overly complicated and error-prone.

Effectively, with the more common configuration, my mail arrives again.

So, **here is the question**: Is there a path-setting pointing the gem-utility to the location of the sought certificates and if so, where? I do not find any configuration file of that type and do not remember to have ever modified one for the gem-utility.
